src/AggregateFunctions/AggregateFunctionGroupArraySorted.h

```diff
@@ -0,0 +1,321 @@
+#pragma once
+
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnVector.h>
+#include <AggregateFunctions/AggregateFunctionGroupArray.h>
+#include <Functions/array/arraySort.h>
+
+#include <Common/Exception.h>
+#include <Common/ArenaAllocator.h>
+#include <Common/assert_cast.h>
+#include <Columns/ColumnConst.h>
+#include <DataTypes/IDataType.h>
+#include <base/sort.h>
+#include <Columns/IColumn.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Common/RadixSort.h>
+#include <algorithm>
+#include <type_traits>
+#include <utility>
+
+
+namespace DB
+{
+struct Settings;
+
+namespace ErrorCodes
+{
+    extern const int TOO_LARGE_ARRAY_SIZE;
+    extern const int INCORRECT_DATA;
+}
+
+template <typename T>
+struct GroupArraySortedNumericData;
+
+template <typename T>
+struct GroupArraySortedNumericData
+{
+    /// For easy serialization.
+    static_assert(std::has_unique_object_representations_v<T> || std::is_floating_point_v<T>);
+
+    // Switch to ordinary Allocator after 4096 bytes to avoid fragmentation and trash in Arena
+    using Allocator = MixedAlignedArenaAllocator<alignof(T), 4096>;
+    using Array = PODArray<T, 32, Allocator>;
+
+    Array value;
+};
+
+template <typename T, typename Trait>
+class GroupArraySortedNumericImpl final
+    : public IAggregateFunctionDataHelper<GroupArraySortedNumericData<T>, GroupArraySortedNumericImpl<T, Trait>>
+{
+    using Data = GroupArraySortedNumericData<T>;
+    static constexpr bool limit_num_elems = Trait::has_limit;
+    UInt64 max_elems;
+    SerializationPtr serialization;
+
+public:
+    explicit GroupArraySortedNumericImpl(
+        const DataTypePtr & data_type_, const Array & parameters_, UInt64 max_elems_ = std::numeric_limits<UInt64>::max())
+        : IAggregateFunctionDataHelper<GroupArraySortedNumericData<T>, GroupArraySortedNumericImpl<T, Trait>>(
+            {data_type_}, parameters_, std::make_shared<DataTypeArray>(data_type_))
+        , max_elems(max_elems_)
+        , serialization(data_type_->getDefaultSerialization())
+    {
+    }
+
+    String getName() const override { return "groupArraySorted"; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    {
+        const auto & row_value = assert_cast<const ColumnVector<T> &>(*columns[0]).getData()[row_num];
+        auto & cur_elems = this->data(place);
+
+        cur_elems.value.push_back(row_value, arena);
+
+        /// To optimize, we sort (2 * max_size) elements of input array over and over again
```

Comment:
This seems like an optimization that should be done by implementing the `addBatch*` functions so you don't copy values unless you really need it: You get a batch, you get to top `max_elems` elements of that batch, you mixed the current element with those top_elems.

That way you don't need to copying and sorting things unnecessarily.